flux-middle/src/ty/conv.rs

```diff
@@ -96,16 +97,49 @@ impl<'a, 'genv, 'tcx> ConvCtxt<'a, 'genv, 'tcx> {
         ty::Binders::new(ty::FnSig::new(requires, args, ret, ensures), params)
     }
 
+    pub(crate) fn conv_enum_def(
+        genv: &mut GlobalEnv,
+        enum_def: core::EnumDef,
+    ) -> Option<Vec<PolyVariant>> {
+        let mut cx = ConvCtxt::new(genv);
+        let variants: Vec<PolyVariant> = enum_def
+            .variants
+            .into_iter()
+            .map(|variant| cx.conv_variant(variant))
+            .collect();
+
+        // Return `None` if there are *no* refined variants as, in that case,
+        // we want to fall back to using "default" `rustc` variant-signatures
+        // at the match/downcast sites as done in `GlobalEnv::variant`.
+        if variants.is_empty() {
+            None
+        } else {
+            Some(variants)
+        }
+    }
+
+    fn conv_variant(&mut self, variant: core::VariantDef) -> PolyVariant {
+        let sorts = self.conv_params(&variant.params);
+        let fields = variant
+            .fields
+            .iter()
+            .map(|ty| self.conv_ty(ty, 1))
+            .collect_vec();
+        let ret = self.conv_ty(&variant.ret, 1);
+        let variant = ty::VariantDef::new(fields, ret);
+        Binders::new(variant, sorts)
+    }
+
     pub(crate) fn conv_struct_def(
         genv: &GlobalEnv,
         struct_def: &core::StructDef,
-    ) -> Option<ty::VariantDef> {
+    ) -> (ty::AdtDef, Option<ty::PolyVariant>) {
```

Comment:
Since we removed the insertion of AdtDef in register_struct we don't have to return the AdtDef here.